feat: Added keyboard navigation to Compose Actions Menu

mananjadhav opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details
<!-- Explanation of the change or anything fishy that is going on -->
- Added keyboard navigation and keyboard based menu trigger for ~Compose Actions Menu~ `<PopoverMenu />`
- Updated changes consists of using `ArrowKeyFocusManager` and `KeyboardShortcuts` lib for keyboard handling.
- **Impacted Components:**
        -   `ReportActionComposer`, additional actions menu consisting `Request Money`, `Split Money` and `Add Attachment`
        -  `FAB Menu/Sidebar Screen`, which pops up when clicking the green floating action Menu
        - `AvatarImagePicker`, which can be tested with Profile Picture
        - `Payments List`, which pops up when clicking Add Payment Methods
        - `ButtonWithMenu` component, which is available when we select the Payment method when settling IOU
        - `ThreeDotsMenu /Workspace Menu`, which is visible on the workspace RHS header

### Fixed Issues
<!---
Please replace GH_LINK with the link to the GitHub issue this Pull Request is fixing.
Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the issue; otherwise, the linking will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<number-of-the-issue>

Do NOT only link the issue number like this: $ #<number-of-the-issue>
--->
$ https://github.com/Expensify/App/issues/7317

### Tests
1. Tested keyboard navigation with arrow keys
2. Tested each menu trigger by clicking enter
3. Checked the focus border blue color shows for active menu
4. Tested the components mentioned in the GH body "Impacted Components" section.

<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

- [x] Verify that no errors appear in the JS console

### QA Steps
**Web and Desktop**

**A. Report Action Composer**
1. With the cursor in compose box Type **Shift+Tab** to navigate to ➕
2. Click on the ➕ icon to open the compose menu
3. With the Up and Down Arrow keys, track the movement of the selected menu item. Selected Menu item should be in grey background.
4. On selecting a particular menu item and pressing `Enter` key it should open the respective menu. 
       a. Request Money, it should Request Money RHS
       b. Send Money, should open Send Money RHS
       c. Add attachment, should open File Picker
5. Escape should hide the Popover

**B. Payment Method List**
1. Open the RHS by clicking the profile picture
2. Click on the Payments Menu
3. Click on the "Add Payment method"
4. With the Up and Down Arrow keys, track the movement of the selected menu item. Selected Menu item should be in grey background.
5. On selecting a particular menu item and pressing `Enter` key it should open the respective menu. 
      a. Debit Card, should open the Add Debit card page
      b. Bank Account, should open the Plaid popup
6.  Escape should hide the Popover


**C. Sidebar Screen/FAB Menu**
1. On the Home Screen, click on the Green FAB (floating action button)
2. With the Up and Down Arrow keys, track the movement of the selected menu item. Selected Menu item should be in grey background.
3. On selecting a particular menu item and pressing `Enter` key it should open the respective menu. 
       a. New Chat, should open New chat search screen
       b. New Group, should open Search Member selection screen
       c. Send Money, should open IOU Amount screen, with title "Send Money"
       d. Request Money, should open IOU Amount screen, with title "Request Money"
       c. Split Bill, should open IOU Amount screen, with title "Split Bill"
4. Escape should hide the Popover

**D. Button with Menu (IOU Payments Settlement options)**
1. Find any chat with outstanding IOU
2. Click on the amount in the LHS, it should open up the IOU Transactions screen
3. At the bottom you should see Settlement Button, click on the down arrow button on the right 
4. With the Up and Down Arrow keys, track the movement of the selected menu item. Selected Menu item should be in grey background.
5. Selecting the option from the menu, should set the title of the green settlement button to the selected menu item text
6. Escape should hide the Popover

**E. Avatar Picker / Profile Picture**
1. Open the Settings menu by clicking Profile Picture
2. Click on the Avatar on the RHS
3. Click on the camera icon (edit icon) on the Avatar
4. With the Up and Down Arrow keys, track the movement of the selected menu item. Selected Menu item should be in grey background.
5. If you don't have a profile picture then only one menu option will be available to upload the picture, clicking should open the File Manager
6. If you have a profile picture (upload one if you don't), then you should be able to navigate with arrow keys between "upload photo" and "remove photo"
7. Escape should hide the Popover

**F. Workspace Menu/Three Dots Menu**
1. Open the Settings Menu clicking the Profile Picture
2. Select any workspace on the RHS
3. Open the Three Dots Menu (<img width="16" alt="image" src="https://user-images.githubusercontent.com/3069065/188285245-0969763a-a463-49c8-bf1a-8eff0eb5d533.png">)
4. With the Up and Down Arrow keys, track the movement of the selected menu item. Selected Menu item should be in grey background.
5. On selecting a particular menu item and pressing `Enter` key it should do the following:
        a. "New Workspace", should create a new Workspace
        b. "Delete Workspace", should open the confirm delete popup.
6.  Escape should hide the Popover
----

**Mobile Web, iOS and Android App**

**A. Report Action Composer**
1. Keyboard shortcuts won't work here, so we click the FAB (➕) button
2. For this we just need to check that clicking each menu item should open its respective menu
3. No crashes or exceptions on these items.
4. Tapping outside the popover should close it.

**B. Payment Method List**
1. Open the RHS by clicking the profile picture
2. Click on the Payments Menu
3. Click on the "Add Payment method"
4. For this we just need to check that clicking each menu item should open its respective menu
5. No crashes or exceptions on these items.
6. Tapping outside the popover should close it.

**C. Sidebar Screen/FAB Menu**
1. On the Home Screen/Chat List screen, click on the Green FAB (floating action button)
2. For this we just need to check that clicking each menu item should open its respective menu
3. No crashes or exceptions on these items.
4. Tapping outside the popover should close it.

**D. Button with Menu (IOU Payments Settlement options)**
1. Find any chat with outstanding IOU
2. Click on the amount in the LHS, it should open up the IOU Transactions screen
3. At the bottom you should see Settlement Button, click on the down arrow button on the right 
4. Selecting the option from the menu, should set the title of the green settlement button to the selected menu item text
5. Tapping outside the popover should close it.

**E. Avatar Picker / Profile Picture**
1. Open the Settings menu by clicking Profile Picture
2. Click on the Avatar on the RHS
3. Click on the camera icon (edit icon) on the Avatar
4. If you don't have a profile picture then only one menu option will be available to upload the picture, clicking should open the File Manager
5. If you have a profile picture (upload one if you don't), then you should be able to navigate with arrow keys between "upload photo" and "remove photo"
6. Tapping outside the popover should close it.

**F. Workspace Menu/Three Dots Menu**
1. Open the Settings Menu clicking the Profile Picture
2. Select any workspace on the RHS
3. Open the Three Dots Menu (<img width="16" alt="image" src="https://user-images.githubusercontent.com/3069065/188285245-0969763a-a463-49c8-bf1a-8eff0eb5d533.png">)
4. On selecting a particular menu item and pressing `Enter` key it should do the following:
        a. "New Workspace", should create a new Workspace
        b. "Delete Workspace", should open the confirm delete popup.
5.  Tapping outside the popover should close it.

<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
13. Upload an image via copy paste
14. Verify a modal appears displaying a preview of that image
--->

- [x] Verify that no errors appear in the JS console

### PR Review Checklist
<!--
This is a checklist for PR authors & reviewers. Please make sure to complete all tasks and check them off once you do, or else Expensify has the right not to merge your PR!
-->
#### Contributor (PR Author) Checklist
- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] iOS / native
    - [x] Android / native
    - [x] iOS / Safari
    - [x] Android / Chrome
    - [x] MacOS / Chrome
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] Any functional components have the `displayName` property
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.


### Screenshots
<!-- Add screenshots for all platforms tested. Pull requests won't be merged unless the screenshots show the app was tested on all platforms.-->

#### Web
<!-- Insert screenshots of your changes on the web platform-->
**Composer Menu**

https://user-images.githubusercontent.com/3069065/186272807-403940e8-bc8c-4b92-9d13-f5c41f51ae0c.mp4

**Sidebar Screen**

https://user-images.githubusercontent.com/3069065/188283702-73ced320-9faf-4115-957f-d226599d77e0.mp4


**Avatar Image Picker**

https://user-images.githubusercontent.com/3069065/188283684-e8a62f37-54ff-449e-94ce-df45c4b39eac.mp4

**Button with Menu**

https://user-images.githubusercontent.com/3069065/188283691-45f3461e-f2bb-4dd2-aea3-8d2dc490bf77.mp4

**Payments Menu**

https://user-images.githubusercontent.com/3069065/188283694-59b70e4c-fae9-4282-935c-7219e2ffbaba.mp4

**Workspace Menu/Three Dots Menu**

https://user-images.githubusercontent.com/3069065/188437744-cfba9931-08f2-4c7d-a593-1c901075c223.mp4


#### Mobile Web
<!-- Insert screenshots of your changes on the web platform (from a mobile browser)-->
https://user-images.githubusercontent.com/3069065/186273977-21714e30-253a-49d5-929e-82d5b79b0540.mp4

#### Desktop
<!-- Insert screenshots of your changes on the desktop platform-->
**Composer Menu**

https://user-images.githubusercontent.com/3069065/186272851-35d169a4-c600-491d-a943-8ab43bcd1707.mp4

**All Components (mentioned in the Web individually)**

https://user-images.githubusercontent.com/3069065/188283800-488ef8e5-78ee-46c0-a6f2-73fc2c213b2a.mp4

**Workspace/ThreeDots Menu (revised)**

https://user-images.githubusercontent.com/3069065/188438395-c8ef1d82-bfe4-4472-85e3-466161d3307a.mp4

#### iOS
<!-- Insert screenshots of your changes on the iOS platform-->
https://user-images.githubusercontent.com/3069065/186274361-1fe2dded-c530-46c7-bac7-536a118fe004.mp4


#### Android
<!-- Insert screenshots of your changes on the Android platform-->
https://user-images.githubusercontent.com/3069065/186273654-6656e9b6-2155-47b4-b229-281557d7c504.mp4

mananjadhav committed:
feat: Add keyboard nav prop

mananjadhav committed:
feat: MenuItem focused handling

mananjadhav committed:
feat: Added keyboard nav handling to BasePopover

mananjadhav committed:
feat: Force disabled keyboard nav to native popover

mananjadhav committed:
feat: Added border color focus and nav to compose actions

mananjadhav committed:
feat: Added separate styles for menu Item

mananjadhav committed:
feat: Reset activeMenuIndex on Popver close

mananjadhav commented:
@parasharrajat @puneetlath PR Raised.

Questions:
1. The QA steps in the linked issue says "Cmd+Tab", I am guessing they mean "Shift+Tab" and not Command. 
2. I've kept the border radius on focus as 0 for the blue border.
3. For a single menu item, it doesn't do anything, just shows a blue border

parasharrajat commented on src/components/PopoverMenu/BasePopoverMenu.js:
```diff
@@ -26,13 +26,88 @@ const defaultProps = {
+                this.props.onItemSelected(this.props.menuItems[this.state.activeMenuIndex]);
+            }
+        };
+        document.addEventListener('keydown', this.keyDownHandler, true);
```
Use KeyboardShorcut lib here.

mananjadhav replied:
Umm a little confused, how `KeyboardShortcut` lib will work here? I thought `KeyboardShortcuts` is to bind a shortcut to the action and subscribe, and here we need for navigation. Can you throw some light?

parasharrajat replied:
Of yeah Sorry This is fine..

roryabraham replied:
I'm confused why we couldn't or wouldn't use `KeyboardShortcut` for this PR...

parasharrajat reviewed (changes requested):
Tab focus has this UI 
![image](https://user-images.githubusercontent.com/24370807/154848467-b4893c85-143a-4eb9-87dc-bcbfcb3bc89f.png)

And when we hover the blue border hides.

mananjadhav commented:
> Tab focus has this UI

Hmm. I checked Arrow keys, not Tab. So should we also change `activeMenuIndex` based on Tab movement?

parasharrajat commented:
Tab movement automatically works. You just need to adjust the Border UI> 
1. Shift+tab. 
2. Tab.
3. Arrow keys 

All should focus on the item. Which it does but UI needs to be adjusted. @mananjadhav could you please record a video for all in the above list. So that we can assess UI. cc: @Expensify/design

mananjadhav commented:
> You just need to adjust the Border UI

Done

> Tab movement automatically works.

It works but is not in sync with Arrow keys. If I am at the second menu item,  pressing the Tab should take me to the third menu item, it doesn't. What I'll have to do is update `activeMenuIndex` with Tab and Shift+Tab as well. Here's the video of the problem.

https://user-images.githubusercontent.com/3069065/154851976-bb20d811-b862-4eb0-bbd2-f60f9757c0a5.mp4

parasharrajat commented:
Yeah, That as well. Thanks for bringing this up. You can use the OnFocus event. That should suffice.

mananjadhav committed:
feat: Removed side borders

parasharrajat commented on src/components/PopoverMenu/BasePopoverMenu.js:
```diff
@@ -26,13 +26,88 @@ const defaultProps = {
+    }
+
+    setupEventHandlers() {
+        if (!document) {
```
NAB: Although this is a safe trick to manage the native platform we don't do this for other components. I leave this to other reviewer

mananjadhav replied:
This was done in EmojiPicker and hence I used it here..

parasharrajat replied:
on EmojiPicker this is done on the index.js file. But here this code would run on all platforms. Using `Document` global on native is bad but this check prevents the crash. But this could be an antipattern of our codebase. I am leaving this to @puneetlath .

mananjadhav replied:
Because we had to maintain the `activeIndex` state, and reduce the code I put the check.

I'll update it into platform-specific code.

parasharrajat commented on src/components/PopoverMenu/index.native.js:
```diff
@@ -28,6 +28,7 @@ class PopoverMenu extends Component {
             <BasePopoverMenu
                 // eslint-disable-next-line react/jsx-props-no-spreading
                 {...this.props}
+                allowKeyboardNavigation={false}
```
This makes the code harder to read. This is a very bad example of using props. 
You are setting the same prop to true on native and false on the web. Although this should never work on natives.
```suggestion
```

mananjadhav replied:
I didn't understand. 

> You are setting the same prop to true on native and false on the web

I am setting it false on `index.native.js` and user-driven for `index.js`. This is to ensure that if the developer sets the prop by mistake it overrides it to `false`

parasharrajat replied:
That is misleading. we should not do that. This prop should not have an effect on natives at all even if the developer sets it true on natives.

shawnborton commented:
It looks like tabbing between menu items causes the items to slightly jump around. Could you please take a look at that? Perhaps it's due to the outline style that's being applied, but it looks like we're getting small 1-2px movements. I think the menu should stay perfectly in place and not move around, and you should only see outline styles being applied to the focused item.

parasharrajat commented:
I agree with @shawnborton 's comment. This is happening as the not-focused item does not have any border but focused has which takes 1px on each edge. 

To fix this, I think we should use a transparent border for not-focused state so it does not cause a layout shift.

mananjadhav committed:
Merge branch 'main' of https://github.com/mananjadhav/App into feat/keyboard-navigation-actions-menu

parasharrajat commented:
cc: @mananjadhav :arrow_up:

parasharrajat commented:
@mananjadhav  Do you think we can reuse code from https://github.com/Expensify/App/pull/7702? I was hoping to create a generic utility to manage arrow navigation as this is common to a few components but It seems Rory already created it in the linked PR. We can reuse that in this PR. Let's hold it till that PR is merged.

mananjadhav commented:
Was reviewing. Yeah looks like we can reuse. Lets wait for it to get merged.

parasharrajat commented:
Please as HOLD in the title for clear indication.

parasharrajat commented:
Ok. It seems 7702 is put on hold. At this stage, it would be best to proceed without those changes. later when that PR is merged these changes could be refactored as well.  

@mananjadhav you are good to move ahead with the requested [changes](https://github.com/Expensify/App/pull/7846#issuecomment-1046643106).

mountiny reviewed (commented):
Updated the title, not on hold anymore.

mananjadhav committed:
Merge branch 'main' of https://github.com/mananjadhav/App into feat/keyboard-navigation-actions-menu

mananjadhav committed:
fix: removed keyboard flag from native popover

mananjadhav committed:
feat: added transparent border for non focused items

mananjadhav committed:
feat: add onFocus to MenuItem Pressable

mananjadhav committed:
feat: set activeMenuIndex on tab

mountiny commented:
What is the state of this PR? 🤔 kind off got lost 😄 is it ready for a review now? If yes, we should probably merge main first.

mananjadhav commented:
So sorry @mountiny. I had to extend the leave and I’ve resumed this week. I am closing all my issues one by one and this will also get closed by the end of this week.

mountiny commented:
@mananjadhav no worries, Manan, it is important to take the time we need sometimes. Thanks for the update and take it step by step, dont overwork yourself now 😅

roryabraham commented on src/components/PopoverMenu/BasePopoverMenu.js:
```diff
@@ -26,13 +26,88 @@ const defaultProps = {
+        document.addEventListener('keydown', this.keyDownHandler, true);
+    }
+
+    highlightActiveMenu(arrowKey) {
```
Missing JSDoc

mananjadhav replied:
Keyboard shortcuts are more to do with hotkeys. This was more of arrow navigation, hence used this

roryabraham reviewed (changes requested):
I wonder if this PR could use the `ArrowKeyFocusManager` I've been working on in https://github.com/Expensify/App/pull/7702? I am hoping it's almost ready and I think it could DRY up a lot of this code.

mananjadhav commented:
@roryabraham We did put this PR on hold for `ArrowKeyFocusManager` #7702 , but I guess it was going to take longer and we thought we'll close this first.  Do you think it'll get merged soon, if yes then it definitely makes sense.

mananjadhav commented:
@roryabraham the other PR is still open. Should we hold this for `ArrowKeyFocusManager` change to be available?

roryabraham commented:
Yes, sorry for the delay and thanks for your patience. Let's wait for https://github.com/Expensify/App/pull/7702 to be ready before continuing with this please.

mananjadhav committed:
Merge branch 'main' of github.com:mananjadhav/App into feat/keyboard-navigation-actions-menu

# Conflicts:
#	src/pages/home/report/ReportActionCompose.js

mananjadhav committed:
Merge branch 'main' of github.com:mananjadhav/App into feat/keyboard-navigation-actions-menu

# Conflicts:
#	src/pages/home/report/ReportActionCompose.js

mananjadhav committed:
Merge branch 'main' of github.com:mananjadhav/App into feat/keyboard-navigation-actions-menu

mananjadhav committed:
Merge branch 'feat/keyboard-navigation-actions-menu' of github.com:mananjadhav/App into feat/keyboard-navigation-actions-menu

parasharrajat commented:
Let's take this off hold.

mananjadhav committed:
fix: rollback basepopover changes

mananjadhav committed:
Merge branch 'main' of github.com:mananjadhav/App into feat/keyboard-navigation-actions-menu

# Conflicts:
#	src/pages/home/report/ReportActionCompose.js

mananjadhav committed:
Merge branch 'main' of github.com:mananjadhav/App into feat/keyboard-navigation-actions-menu

mananjadhav committed:
style: indent and space fixes

roryabraham commented:
@mananjadhav feel free to explicitly re-request reviews once this is ready

mananjadhav committed:
Merge branch 'main' of github.com:mananjadhav/App into feat/keyboard-navigation-actions-menu

mananjadhav committed:
Merge branch 'main' of github.com:mananjadhav/App into feat/keyboard-navigation-actions-menu

mananjadhav committed:
Merge branch 'main' of github.com:mananjadhav/App into feat/keyboard-navigation-actions-menu

mananjadhav committed:
Merge branch 'main' of github.com:mananjadhav/App into feat/keyboard-navigation-actions-menu

# Conflicts:
#	src/components/MenuItem.js

mananjadhav committed:
fix: remove unwanted prop

mananjadhav committed:
feat: added ArrowKeyFocusManager for MenuItems

mananjadhav committed:
feat: focus when the index match

mananjadhav committed:
refactor: remove unused onfocus prop

mananjadhav committed:
feat: add enter handler

mananjadhav committed:
feat: attach listener only to isVisible popup

mananjadhav committed:
fix: reset index on closure and other scenarios

mananjadhav committed:
fix: correct function name

mananjadhav committed:
fix: added method binding

mananjadhav committed:
fix: remove remove key function on modal hide

mananjadhav committed:
fix: remove remove key function on modal hide

mananjadhav committed:
Merge branch 'feat/keyboard-navigation-actions-menu' of github.com:mananjadhav/App into feat/keyboard-navigation-actions-menu

# Conflicts:
#	src/components/PopoverMenu/BasePopoverMenu.js

mananjadhav commented:
I can see that the other places we're not syncing the Tab focus and the Arrow focuses, hence I've removed that part of the code and updated the PR. It now uses `ArrowKeyFocusManager`.

roryabraham commented on src/components/PopoverMenu/BasePopoverMenu.js:
```diff
@@ -26,13 +29,64 @@ const defaultProps = {
+        this.removeKeyboardShorcuts();
+    }
+
+    onModalHide() {
```
Make sure methods are named for what they do, not what they handle

mananjadhav replied:
Moved the method as inline.

roryabraham commented on src/components/PopoverMenu/BasePopoverMenu.js:
```diff
@@ -48,17 +102,24 @@ class BasePopoverMenu extends PureComponent {
+                                title={item.text}
+                                description={item.description}
+                                onPress={() => this.props.onItemSelected(item)}
+                                focused={this.state.focusedIndex === menuIndex}
```
This prop should be `isFocused` because it's a boolean

mananjadhav replied:
It wasn't added in this PR. I am using an existing prop. Do you want me to refactor it for `MenuItem`. Might have to check the places where it'll have to be updated.

roryabraham commented on src/components/PopoverMenu/popoverMenuPropTypes.js:
```diff
@@ -49,13 +49,17 @@ const propTypes = {
     disableAnimation: PropTypes.bool,
+
+    /** Whether to allow arrow key actions on the list */
+    enableArrowKeysActions: PropTypes.bool,
```
```suggestion
    shouldEnableArrowKeysActions: PropTypes.bool,
```

mananjadhav replied:
Updated.

roryabraham commented on src/components/PopoverMenu/BasePopoverMenu.js:
```diff
@@ -26,13 +29,64 @@ const defaultProps = {
+        this.props.onMenuHide();
+    }
+
+    removeKeyboardShorcuts() {
```
I think that this method isn't needed. It's only used once so just inline it in `componentWillUnmount`

mananjadhav replied:
Removed.

mananjadhav committed:
refactor: remove extra methods

mananjadhav committed:
refactor: rename the flag to enable key actions

mountiny reviewed (commented):
Looks good to me, @parasharrajat would you be able to give this a look and some testing? Thank you very much 🙇

mananjadhav commented on src/components/PopoverMenu/BasePopoverMenu.js:
```diff
@@ -26,13 +29,56 @@ const defaultProps = {
+            return;
+        }
+
+        const shortcutConfig = CONST.KEYBOARD_SHORTCUTS.ENTER;
+        this.unsubscribeEnterKey = KeyboardShortcut.subscribe(shortcutConfig.shortcutKey, () => {
+            if (this.state.focusedIndex === -1) {
+                return;
+            }
+            this.props.onItemSelected(this.props.menuItems[this.state.focusedIndex]);
+            this.updateFocusedIndex(-1); // Reset the focusedIndex on selecting any menu
+        }, shortcutConfig.descriptionKey, shortcutConfig.modifiers, true);
+    }
```
Are the arrow functions not working or just the enter?

mountiny replied:
> Let's try to wrap this one quickly. It is close to the end. Though, no rush.

😂 

Clear instructions

marcaaron replied:
am dying 😂

mananjadhav replied:
Yeah I am debugging whats wrong. I'll have it closed by tomorrow.

mananjadhav replied:
Can you check now @parasharrajat ?

mananjadhav committed:
refactor: moved inline function to class method

mananjadhav commented on src/components/PopoverMenu/BasePopoverMenu.js:
```diff
@@ -26,13 +29,56 @@ const defaultProps = {
+                onModalHide={() => {
+                    this.updateFocusedIndex(-1); // Reset the focusedIndex on modal hide
+                    this.props.onMenuHide();
+                }}
```
Updated

mananjadhav committed:
fix: attach and remove listeners on hide/show

mananjadhav committed:
fix: change maxIndex < 1 check

mananjadhav committed:
fix: condition to remove multiple attacher listener

mananjadhav commented on src/components/ArrowKeyFocusManager.js:
```diff
@@ -26,7 +26,7 @@ class ArrowKeyFocusManager extends Component {
         const arrowDownConfig = CONST.KEYBOARD_SHORTCUTS.ARROW_DOWN;
 
         this.unsubscribeArrowUpKey = KeyboardShortcut.subscribe(arrowUpConfig.shortcutKey, () => {
-            if (this.props.maxIndex < 1) {
```
@roryabraham I need help in confirming this check. Any specific reason we used 1 earlier? It causes an issue when we've got only one element in the list. Hence I've changed to 0.

roryabraham replied:
I am pretty skeptical of this change – there are a lot of regression tests that need to be done locally for this PR. I am going to post them in the issue description.

roryabraham replied:
Sorry, maybe it's just my eyes, but to me none of those items look preselected / highlighted.

My knee-jerk reaction here was that on web + desktop the first item should be highlighted, but on iOS/Android/mWeb probably not. After thinking about it some more though, this is a power-user feature and therefore none of the options need to be pre-selected on any platform, because doesn't need to be immediately obvious that you can use the keyboard to navigate these list items.

So here's what I think we should do for the expected behavior:

- When you open the menu, nothing is highlighted.
- If you press ArrowUp at this point, the bottom item will be highlighted. If you press ArrowDown, the top item will be highlighted.
- After an option is highlighted, the list should cycle such that:
    - If the top item is highlighted and you press ArrowUp, then the bottom item is highlighted
    - The the bottom item is highlighted, and you press ArrowDown, then the top item is highlighted
    
Does that sound good? If not, let's chat in slack about it. Thanks!

mananjadhav replied:
@roryabraham To fix that the it should allow you to select the when the there's only 1 item in the list, I updated the check for `maxIndex < 0`.

roryabraham commented:
Sorry for the format, I've asked about getting you all access to TestRail [here](https://expensify.slack.com/archives/C02NK2DQWUX/p1661765427354489)

<img width="1512" alt="image" src="https://user-images.githubusercontent.com/47436092/187171174-db8d1bd8-82b4-41de-8e0b-ef0913bbf41c.png">
<img width="1520" alt="image" src="https://user-images.githubusercontent.com/47436092/187171246-99de16a0-19c1-4cdd-9ec9-033b85604290.png">


<img width="1508" alt="image" src="https://user-images.githubusercontent.com/47436092/187171611-4f9f8cf6-420f-4f24-9537-0b86cf27a587.png">

mananjadhav commented on src/components/PopoverMenu/BasePopoverMenu.js:
```diff
@@ -26,13 +29,75 @@ const defaultProps = {
+    }
+
+    attachKeyboardListener() {
+        if (!this.props.shouldEnableArrowKeysActions || !this.props.isVisible || this.unsubscribeEnterKey) {
+            return;
+        }
```
I'll update.

mananjadhav commented:
@roryabraham I read the chat but is there a way to get a access to the limited testrail cases? to test this issue.

mananjadhav committed:
Merge branch 'main' of github.com:mananjadhav/App into feat/keyboard-navigation-actions-menu

mananjadhav commented:
@roryabraham Anyway to get a text dump? so that I can probably just create a checklist and update here with the testing.

mananjadhav committed:
refactor: changed conditions for keyboard listeners

roryabraham commented:
Here you go:

#### Search Page
##### All Platforms

1. Click/tap on the magnifying glass icon to open the search page.
2. Click/tap on the x icon to close the search page.
3. Click/tap on the magnifying glass to open the search page.
4. Ensure you can scroll up and down in the results.
5. Enter a search query – the list should update accordingly.
6. Click/tap on a result: the search page should close and that chat should open.

##### Web/Desktop

1. Press CMD+K to open the search page.
1. Press Esc to close the search page.
1. Press CMD+K to reopen the search page.
1. Type in a search query.
1. Use the arrow keys to go through the search results. Make sure that the list "wraps", so:
    - If you are focused on the last result and press ArrowDown you are taken to the first result
    - If you are focused on the first result and press ArrowUp you are taken to the last result
1. With the focus somewhere in the middle of the list, update the search input such that there are no results
1. Press the arrow keys a few times – nothing should happen.
1 Undo the change from the previous step by clearing the search input (press `ALT+Delete`). The focus should be on the first element of the list, and you should be able to use the arrows to navigate focus (as before)
1. Enter a new search term. Verify that the first result in the list is highlighted, and that you can navigate up and down from there using the arrow keys.
1. With a chat highlighted on the search page, press Enter. The search page should close and that chat should open.
1. Use the mouse to click on the magnifying glass icon and reopen the search page.
1. Press Esc to close the search page.
1. Click on the magnifying glass to reopen the search page.
1. Use the arrow keys and the enter key to select and open a different chat.
1. Press `CMD+K` to reopen the search page.
1. Use the arrows keys to highlight a search result, then press `CMD+Enter`. The search page should close and your selected chat should open.

#### New Chat Page
##### All Platforms

1. Click/tap on the green + icon on the home screen
2. Click/tap on New Chat.
3. Ensure that you can scroll the list of results.
4. Enter a search query and the results should update accordingly.
5. click/tap on a search result
6. The New Chat Page should close and a chat with your selected result should open.

##### Web/Desktop

1. Click on the green + icon on the home screen
1. Click on New Chat
1. Press ESC – the New Chat Page should close.
1. Click on the green + icon on the home screen.
1. Click on New Chat
1. Use the arrow keys to traverse the list of results. Make sure that the list "wraps", so:
    - If you are focused on the last result and press ArrowDown you are taken to the first result
    - If you are focused on the first result and press ArrowUp you are taken to the last result
1. With the focus somewhere in the middle of the list, update the search input such that there are no results.
1. Press the arrow keys a few times – nothing should happen.
1. Undo the change from the previous step – the focus should be back on the first element in the list
1. With a result highlighted, press Enter and that chat should open.
1. Reopen the New Chat page and use the arrow keys to highlight a different chat than the one you have open.
1. Press CMD+Enter and that chat should open.

#### New Group Page
#####All Platforms

1. Click/tap on the green + icon on the home screen
2. Click/tap on New Group.
3. Ensure that you can scroll the list of results.
4. Enter a search query and the results should update accordingly.
5. Click/tap on a result. That result should become checked and move to the top of the results. The list should scroll to the top automatically and you should be able to see your new selection.
6. Click/tap on another result. That result should become checked and move to the top of the results. The list should scroll to the top automatically and you should be able to see your new selection.
7. Click/tap on Create Group. The New Group page should close and a chat with all the selected users should open.

#### Web/Desktop

1. Click on the green + icon on the home screen.
1. Click New Group. The New Group Page should open.
1. Press ESC. The New Group page should close.
1. Press CMD+SHIFT+K. The New Group Page should open.
1. Press ESC. The New Group page should close.
1. Reopen the New Group Page.
1. If necessary, repeat the following until there are enough elements in the New Group page that the results are scrollable:
    1. Close the New Group page.
    1. Open the New Chat page.
    1. Create a chat with a user that you've not interacted with before.
    1. Reopen the New Group page and check if it's scrollable. If not, go back to the first step.
1. Use the arrow keys to traverse the list of results. Make sure that the list "wraps". There is a notable difference in expected behavior between this page and the New Chat Page in that the search bar is one of the focusable elements.
    - If you are focused on the last result and press ArrowDown you are taken to the search bar. None of the elements should be highlighted.
    - If you are focused on the first result and press ArrowUp you are taken to the search bar. None of the elements should be highlighted.
    - If you are focused on the search bar and press ArrowUp, you are taken to the last result.
    - If you are focused on the search bar and press ArrowDown, you are taken to the first result.
1. With one of the search results highlighted, press Enter. That result should get checked and move to the top of the list. The list should scroll to the top automatically and you should be able to see your new selection.
1. Use the ArrowDown key to highlight the first unselected result. It should be completely visible.
1. Use the arrow keys and Enter to select multiple results.
1. Use the arrow keys to highlight a selected result.
1. Press Enter to de-select that result.
1. With a result highlighted, press CMD+Enter, and a group should be opened with your selected results.
1. Reopen the New Group Page
1. Without selecting any results, press CMD+Enter. Nothing should happen
1. Select several results.
1. Use the arrow keys to go to the search bar (no results highlighted).
1. Press CMD+Enter, and a group should be opened with your selected results.
1. Reopen the New Group Page and select several results.
1. Use the arrow keys to go to the search bar (no results highlighted).
1. Press Enter (not CMD+Enter), and a group should be opened with your selected results.

#### IOU Participants (split bill)
##### All Platforms

1. Click/tap on the green + icon on the home screen.
2. Click/tap on Split Bill
3. Enter an amount.
4. Ensure that you can scroll the list of results.
5. Enter a search query and the results should update accordingly.
6. Make sure you can toggle / untoggle list items.
7. With some items selected and others not, press Next.
8. Verify that you the selected items are shown on the next page.

##### Web/Desktop

1. Click/tap on the green + icon on the home screen.
1. Click/tap on Split Bill
1. Enter an amount and press Next.
1. Enter a search query to get a reasonable-length list of options.
1. Use the arrow keys to traverse the list of results. Make sure that the list "wraps":
    - If you are focused on the last result and press ArrowDown you are taken to the search bar. None of the elements should be highlighted.
    - If you are focused on the first result and press ArrowUp you are taken to the search bar. None of the elements should be highlighted.
    - If you are focused on the search bar and press ArrowUp, you are taken to the last result.
    - If you are focused on the search bar and press ArrowDown, you are taken to the first result.
1. With one of the search results highlighted, press Enter. That result should get checked and move to the top of the list. The list should scroll to the top automatically and you should be able to see your new selection.
1. Use the arrow keys and Enter to select multiple results.
1. Use the arrow keys to highlight a selected result.
1. Press Enter to de-select that result.
1. With a result highlighted, press CMD+Enter, and you should move to the next page with your selected results.
1. Press the back arrow to go to the previous page.
1. With the search bar focused (no results highlighted), press CMD+Enter, and you should move to the next page with your selected results.
1. Press the back arrow to go to the previous page.
1. With the search bar focused (no results highlighted), press Enter (not CMD+Enter), and you should move to the next page with your selected results.

#### IOU Participants (1:1 request)
##### All Platforms

1. Click/tap on the green + icon on the home screen.
2. Click/tap on Request Money
3. Enter an amount and press Next
4. Ensure that you can scroll the list of results.
5. Enter a search query and the results should update accordingly.
6. Click/tap on a result, and there should be a new IOU request between you and your selected user.

##### Web/Desktop

1. Click/tap on the green + icon on the home screen.
1. Click/tap on Request Money
1. Enter an amount and press Next
1. Enter a search query to get a reasonable-length list of results.
1. Use the arrow keys to traverse the list of results. Make sure that the list "wraps":
    - If you are focused on the last result and press ArrowDown you are taken to the first result
    - If you are focused on the first result and press ArrowUp you are taken to the last result
1. With a result highlighted, press Enter. That should take you to the next page with your selected user.
1. Click on the back button.
1. Use the arrow keys to highlight different result.
1. Press CMD+Enter. That should take you to the next page with your new selected user.

#### IOU Confirmation List
##### All Platforms

1. Click/tap on the green + icon on the home screen.
2. Click/tap on Split Bill.
3. Enter an amount and press Next.
4. Select some participants and press Next.
5. Make sure that you can not select or deselect participants.
6. Hover the mouse over your participants – it should not change into a pointer cursor.
7. Press Split YOUR_AMOUNT
8. A new request should be created between you and each of your selected bill split participants.
9. Open a group chat.
10. Press the + in the composer and select Split Bill.
11. Make sure you can select and deselect participants, and that their proportional amount adjusts accordingly.
12. Hover your mouse over the participants – it should change into a pointer cursor and clicking on participants should select/deselect them.
13. With some participants selected and others not, press Split YOUR_AMOUNT.
14. A new request should be created between you and each of your selected bill split participants.

##### Web/Desktop

1. Open a group chat.
1. Click/tap on the + in the composer.
1. Click/tap on Split Bill.
1. Enter an amount and press Next.
1. Use the arrow keys to highlight participants. Make sure that the list "wraps":
    - If you are focused on the last result and press ArrowDown you are taken to the text input. None of the elements should be highlighted.
    - If you are focused on the first result and press ArrowUp you are taken to the text input. None of the elements should be highlighted.
    - If you are focused on the text input and press ArrowUp, you are taken to the last result.
    - If you are focused on the text input and press ArrowDown, you are taken to the first result.
1. With a result highlighted, pressing Enter should select and deselect participants, and their proportional amount should adjust accordingly.
1. Deselect all participants.
1. Press CMD + Enter. Verify that nothing happens (no IOU should be created with just yourself)
1. Reselect one or more participants.
1. With no result highlighted, press Enter. A new request should be created between you and each of your selected participants.
1. Open a group chat.
1. Click/tap on the + in the composer.
1. Click/tap on Split Bill.
1. Enter an amount and press Next.
1. With a result highlighted, press CMD+Enter. A new request should be created between you and each of your selected participants.
1. Open a 1:1 DM with another user (not Chronos or Concierge).
1. Press the + button in the compose bar and click Request Money
1. Enter an amount and press Next
1. Press Enter. The IOU should be created.

#### Workspace invite page
##### All Platforms

1. Create a workspace if you do not already have one.
2. Navigate to Settings -> Your Workspace -> Manage Members -> Invite.
3. Verify that you can scroll + search the results on the invite page.
4. Select several users from the list.
5. Press Invite, and verify that you are taken to the previous page and the new invitees are shown as members.

##### Web/Desktop

1. Create a workspace if you do not already have one.
1. Navigate to Settings -> Your Workspace -> Manage Members -> Invite.
1. Use the arrow keys to traverse the list. Verify that the list "wraps":
    - If you are focused on the last result and press ArrowDown you are taken to the search bar. None of the elements should be highlighted.
    - If you are focused on the first result and press ArrowUp you are taken to the search bar. None of the elements should be highlighted.
    - If you are focused on the search bar and press ArrowUp, you are taken to the last result.
    - If you are focused on the search bar and press ArrowDown, you are taken to the first result.
1. With an option highlighted, press Enter. That option should be selected and moved to the top of the list. The list should scroll to the top automatically and you should be able to see your new selection.
1. Select a few more options.
1. Highlight a selected option and press Enter. That should deselect the option.
1. With a few options selected, move focus to the search bar (no items highlighted), and press Enter. That should take you to the previous page with the new invitees shown as members.
1. Press Invite again.
1. Use the arrow keys to select a few options.
1. With a search result highlighted, press CMD+Enter. That should take you to the previous page with the new invitees shown as members.
1. Press Invite again.
1. Use the arrow keys to select a few options.
1. With the search bar focused, press CMD+Enter. That should take you to the previous page with the new invitees shown as members.

#### IOU currency selection
##### All Platforms

1. Press the big green + to open the global create menu.
2. Select "request money"
3. Tap on the currency symbol ($ in US)
4. Search for another currency, verify the results list updates.
5. Select another currency, verify that the currency selection modal closes and the IOU amount page shows the new symbol for the currency you selected.

##### Web/Desktop

1. Press the big green + to open the global create menu.
1. Select "request money"
1. Click on the currency symbol ($ in US)
1. Search for another currency, verify the results lists updates.
1. Use the arrow key to scroll through the list. Verify that it "wraps" such that if you are at the bottom and press ArrowDown, you are taken to the top, and vice-versa.
1. With an item highlighted, change the search term.
1. Verify that the list scrolls back to the top and that the first item is highlighted
1. With an item highlighted, press the enter key.
1. Verify that the currency selection modal close and the IOU amount page shows the new symbol for the currency you selected.
1. Click on the currency symbol.
1. Use the arrow keys to highlight another option.
1. Press CMD+Enter
1. Verify that the currency selection modal close and the IOU amount page shows the new symbol for the currency you selected.

mananjadhav commented on src/components/PopoverMenu/BasePopoverMenu.js:
```diff
@@ -26,13 +29,75 @@ const defaultProps = {
+            return;
+        }
+        this.unsubscribeEnterKey();
+        this.unsubscribeEnterKey = null;
```
removed

mananjadhav commented:
Half done, I'll finish by tomorrow. Are we okay with all the code changes? Especially this [one](https://github.com/Expensify/App/pull/7846#discussion_r955425072)

mananjadhav commented:
Tested regression suite.

roryabraham commented on src/components/PopoverMenu/popoverMenuPropTypes.js:
```diff
@@ -49,13 +49,17 @@ const propTypes = {
     disableAnimation: PropTypes.bool,
+
+    /** Whether to allow arrow key actions on the list */
+    shouldEnableArrowKeysActions: PropTypes.bool,
```
Wondering why we don't just enable this by default and not introduce this prop ... I get that it would make the change in many more places across the app, but standardizing on arrow navigation everywhere seems like a good idea. Due to the extra QA that would be needed we could double the budget.

I would maybe rather do that than have a number of different PRs just to add this prop into different components, and then at some later date just remove the prop entirely.

roryabraham replied:
Another advantage of this approach is that we can create a single issue to update the regression suite with all the tests needed in the different places where `<PopoverMenu />` is used

mananjadhav replied:
I did think about reversing it by making it true by default and we explicitly disable where not needed. There are about 5-6 places where `<PopoverMenu />` is used. 

> I would rather do that than have a number of different PRs just to add this prop into different components
Good thought.

roryabraham replied:
I looked though the usage and I _think_ we can have keyboard navigation everywhere this component is used. If that's the case we can just remove this prop entirely

mananjadhav replied:
Okay give me a day I'll remove the prop, test again for the other screens. 

> Due to the extra QA that would be needed we could double the budget.

It would be great if this can be done.

mananjadhav replied:
Updated

mananjadhav committed:
Merge branch 'main' of github.com:mananjadhav/App into feat/keyboard-navigation-actions-menu

mananjadhav committed:
refactor: make keyboard actions available by default

mananjadhav commented:
@roryabraham @parasharrajat PR is updated by making the keyboard navigation available by default. Also updated the screencasts and QA steps for the additional components. Please take a look.

-----

I found one issue, which isn't related to my change afaik, when clicking the menu items in the `ThreeDotsMenu`,  the popover doesn't close. What do you think of putting it outside the scope of this PR?

roryabraham commented:
Thanks for adding screenshots and QA steps! However, I'm not able to reproduce the ThreeDotsMenu issue on staging, so I think we need to address it here

mananjadhav commented:
> Thanks for adding screenshots and QA steps! However, I'm not able to reproduce the ThreeDotsMenu issue on staging, so I think we need to address it here

In the current PR, it works fine for all the other components, except for ThreeDotsMenu. Let me check once again. Thanks for pointing it out.

roryabraham commented:
It _looks_ like what's happening is:

1. ThreeDots icon is focused
1. When you open the PopoverMenu, the focus shifts to that
1. You use the arrow keys to select an option from the PopoverMenu.
1. Then you press enter to select the highlighted option from the PopoverMenu
1. This executes the callback and closes the PopoverMenu, and the focus shifts back to the ThreeDots icon. The same `Enter` event fires on the ThreeDots menu and it reopens.

Something like that

mananjadhav committed:
Merge branch 'main' of github.com:mananjadhav/App into feat/keyboard-navigation-actions-menu

mananjadhav committed:
fix: update threedots menu popover visibility logic

mananjadhav commented:
@roryabraham @parasharrajat Fix added for `ThreeDotsMenu`. Ready for review.

roryabraham reviewed (approved):
- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] iOS / native
    - [x] Android / native
    - [x] iOS / Safari
    - [x] Android / Chrome
    - [x] MacOS / Chrome
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] Any functional components have the `displayName` property
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

roryabraham commented:
LGTM and tests well

mountiny reviewed (approved):
Code lgtm! Great job @mananjadhav being persistent with this issue!

roryabraham commented:
Not an emergency, see above comment

roryabraham commented:
Echoing @mountiny's sentiment above, thanks so much for your patience and persistence with this issue. I love how we really applied this solution in a generalized way across the app

parasharrajat reviewed (commented):
Lol, this is merged by my testing is still not completed. The List is very big. But code is looking good.

mananjadhav commented:
> Lol, this is merged by my testing is still not completed

Do ping me if you find anything and I'll try to quickly raise another PR to address any issue/bug.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/3011577367) to staging by @roryabraham in version: 1.1.98-0 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅
